src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java

```diff
@@ -38,30 +39,43 @@ public BufferLeasePool() {
         this.byteBufferSupplier = () -> ByteBuffer.allocateDirect(segmentSize); // TODO configurable extents
         this.queue = new ConcurrentLinkedQueue<>();
         this.bufferLeaseStub = new BufferLeaseStub();
+        this.bufferContainerStub = new BufferContainerStub();
         this.close = new AtomicBoolean();
         this.bufferId = new AtomicLong();
         this.lock = new ReentrantLock();
     }
 
     private BufferLease take() {
         // get or create
-        BufferLease bufferLease = queue.poll();
-        if (bufferLease == null) {
-            bufferLease = new BufferLeaseImpl(bufferId.incrementAndGet(), byteBufferSupplier.get());
+        BufferContainer bufferContainer = queue.poll();
+        BufferLease bufferLease;
+        if (bufferContainer == null) {
+            // if queue is empty or stub object, create a new BufferContainer and BufferLease.
+            bufferLease = new BufferLeaseImpl(
+                    new BufferContainerImpl(bufferId.incrementAndGet(), byteBufferSupplier.get())
+                    ,
+                    this);
+        } else {
+            // otherwise, wrap bufferContainer with phaser decorator (bufferLease)
+            bufferLease = new BufferLeaseImpl(bufferContainer, this);
         }
 
-        bufferLease.addRef(); // all start with one ref
-
         if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug("returning bufferLease id <{}> with refs <{}> at buffer position <{}>", bufferLease.id(), bufferLease.refs(), bufferLease.buffer().position());
+            LOGGER.debug("returning bufferLease id <{}> with refs<{}> at buffer position <{}>", bufferLease.id(), bufferLease.refs(), bufferLease.buffer().position());
+        }
+
+        if (bufferLease.buffer().position() != 0) {
+            LOGGER.error("dirty buffer in pool, terminating");
+            System.exit(1);
```

Comment:
Is hard exiting really wanted here? If so, I'd use enum values for exit like `System.exit(ExitCodes.DIRTY_BUFFER_IN_POOL);` to track what are defined and to catch deal with return codes in shell and so on

Reply:
fix: Add IllegalStateException [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)